drivers/at86rf215/at86rf215_internal.c

```diff
@@ -0,0 +1,192 @@
+{
+    spi_acquire(SPIDEV, CSPIN, SPI_MODE_0, dev->params.spi_clk);
+}
+
+/* only to be used by at86rf215_hardware_reset()
+   can't use normal at86rf215_reg_read() because
+   we already hold the lock */
+static inline uint8_t _get_rf_state_with_lock(at86rf215_t *dev)
+{
+    uint16_t reg = htons(dev->RF->RG_STATE | FLAG_READ);
+    spi_transfer_bytes(SPIDEV, CSPIN, true, &reg, NULL, sizeof(reg));
+    return spi_transfer_byte(SPIDEV, CSPIN, false, 0) & STATE_STATE_MASK;
+}
+
+int at86rf215_hardware_reset(at86rf215_t *dev)
+{
+    /* prevent access during reset */
+    getbus(dev);
+
+    /* trigger hardware reset */
+    gpio_clear(dev->params.reset_pin);
+    xtimer_usleep(AT86RF215_RESET_PULSE_WIDTH);
+    gpio_set(dev->params.reset_pin);
+    xtimer_usleep(AT86RF215_RESET_DELAY);
+
+    unsigned tries = 100;
+    while (_get_rf_state_with_lock(dev) != RF_STATE_TRXOFF &&
+           --tries) {}
+    spi_release(SPIDEV);
+
+    /* bail out if no module is connected */
+    if (tries == 0) {
+        return -ENODEV;
+    }
+
+    /* clear interrupts */
+    at86rf215_reg_read(dev, RG_RF09_IRQS);
+    at86rf215_reg_read(dev, RG_RF24_IRQS);
+    at86rf215_reg_read(dev, RG_BBC0_IRQS);
+    at86rf215_reg_read(dev, RG_BBC1_IRQS);
+
+    return 0;
+}
+
+void at86rf215_reg_write(const at86rf215_t *dev, uint16_t reg, uint8_t value)
+{
+    reg = htons(reg | FLAG_WRITE);
+
+    getbus(dev);
+    spi_transfer_bytes(SPIDEV, CSPIN, true, &reg, NULL, sizeof(reg));
+    spi_transfer_byte(SPIDEV, CSPIN, false, value);
+    spi_release(SPIDEV);
+}
+
+void at86rf215_reg_write_bytes(const at86rf215_t *dev, uint16_t reg, const void *data, size_t len)
+{
+    reg = htons(reg | FLAG_WRITE);
+
+    getbus(dev);
+    spi_transfer_bytes(SPIDEV, CSPIN, true, &reg, NULL, sizeof(reg));
+    spi_transfer_bytes(SPIDEV, CSPIN, false, data, NULL, len);
+    spi_release(SPIDEV);
+}
+
+uint8_t at86rf215_reg_read(const at86rf215_t *dev, uint16_t reg)
+{
+    uint8_t val;
+
+    reg = htons(reg | FLAG_READ);
+
+    getbus(dev);
+    spi_transfer_bytes(SPIDEV, CSPIN, true, &reg, NULL, sizeof(reg));
+    val = spi_transfer_byte(SPIDEV, CSPIN, false, 0);
+    spi_release(SPIDEV);
+
+    return val;
+}
+
+void at86rf215_reg_read_bytes(const at86rf215_t *dev, uint16_t reg, void *data, size_t len)
+{
+    reg = htons(reg | FLAG_READ);
+
+    getbus(dev);
+    spi_transfer_bytes(SPIDEV, CSPIN, true, &reg, NULL, sizeof(reg));
+    spi_transfer_bytes(SPIDEV, CSPIN, false, NULL, data, len);
+    spi_release(SPIDEV);
+}
+
+void at86rf215_filter_ack(at86rf215_t *dev, bool on)
+{
+    /* only listen for ACK frames */
+    uint8_t val = on ? (1 << IEEE802154_FCF_TYPE_ACK)
+                     : (1 << IEEE802154_FCF_TYPE_BEACON)
+                     | (1 << IEEE802154_FCF_TYPE_DATA)
+                     | (1 << IEEE802154_FCF_TYPE_MACCMD);
+
+    at86rf215_reg_write(dev, dev->BBC->RG_AFFTM, val);
+}
+
+void at86rf215_get_random(at86rf215_t *dev, void *data, size_t len)
```

Comment:
I'm not using that, only implemented it because `at86rf2xx` had it too - I'll delete it.

Reply:
Hah, turns out it's used by CSMA after all - I'll look into it.

Reply:
The way the function is used currently, it is only called in state RX.
Since this is an internal function, I think it's save to add this as a pre-condition to avoid unnecessary register reads when servicing the energy detect interrupt.

Reply:
OK!